### ISSUE TITLE
How would I use tsyringe with express?

### ISSUE THREAD

**Issue by emaxedon** [microsoft/tsyringe#41]
I've set up my express project so that I have `controllers`, `services`, and `repositories`.

The following works:
```
// bootstrap.ts
import 'reflect-metadata';
import { container } from 'tsyringe';

import { AuthService, UserService } from './services';
import { UserRepository } from './repositories';
import { AuthController } from './controllers';

container.register('IUserRepository', {
	useClass: UserRepository
});

container.register('IUserService', {
	useClass: UserService
});

container.register('IAuthService', {
	useClass: AuthService
});

export const authController = container.resolve(AuthController);
```

Here `AuthController` depends on `IUserService` and `IAuthService`, and `IUserService` depends on `IUserRepository`.

The problem I am having is when I call a method of `authController` like so:
```
import * as express from 'express';

import { authController } from './bootstrap'

const router = express.Router();

router.get('test', authController.test);
```

If I hit the server's 'test' route, authController.test does get fired, however if inside `authController.test` a dependency is being used the application hangs.

I can use dependencies right after I resolve authController in the bootstrap file, but after the server is started, and when I try to hook up express routes with authController methods the dependencies are "gone".

TLDR;
How do I use `tsyringe` with express where routes fire methods from an `@injectable()` class?

**Comment by emaxedon** [comment]
Another equally strange solution is to do:
```
router.get('test', authController.test.bind(authController));
```
Any idea as to why this works and if there is a better solution overall?

**Comment by mrfsbDev** [comment]
I have the same problem, any other solution?

**Comment by csp04** [comment]
This worked for me.

request.handler.ts
```
import { Request, Response } from 'express';
type RequestHandler = (req: Request, res: Response) => void;
export default RequestHandler;
```

user.controller.ts
```
import { Request, Response } from 'express';
import { inject, singleton } from 'tsyringe';
import UserService from '../services/user.service';
import RequestHandler from './request.handler';

@singleton()
export default class UserController {
  constructor(@inject('UserService') private service: UserService) {}

  getInfo = (): RequestHandler => (req: Request, res: Response) => {
    const params = req.params as { id: string };

    res.send(this.service.getUser());
  };
}
```

user.route.ts
```
...
userRoute.get('/', userController.getInfo());
```
